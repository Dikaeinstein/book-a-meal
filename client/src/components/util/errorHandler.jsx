import React from 'react';
import PropTypes from 'prop-types';
import transformError from '../../helpers/transformError';

/**
 * A reusable error handler HOC
 *
 * @param {JSX} WrappedComponent React component
 * @param {String} defaultMessage Default fallback error message
 * @param {Function} onRetry Retry handler
 * @param {Boolean} retry True or False
 *
 * @returns {JSX} React component with error handling capability
 */
const errorHandler = (WrappedComponent, defaultMessage, onRetry = null, retry = false) => {
  const ErrorWrapper = ({ error, ...rest }) => {
    if (error) {
      return (
        <div className="error-container text-center">
          <h2>
            {transformError(error, defaultMessage)}
          </h2>
          {retry
            ?
              <button
                className="btn btn-default"
                onClick={onRetry}
                style={{ margin: '1rem auto 0 auto' }}
              >
                Retry
              </button>
            : null}
        </div>
      );
    }

    return <WrappedComponent {...rest} />;
  };

  ErrorWrapper.propTypes = {
    /* eslint react/require-default-props: 0 */
    error: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.objectOf(PropTypes.string),
    ]),
  };

  return ErrorWrapper;
};

export default errorHandler;
