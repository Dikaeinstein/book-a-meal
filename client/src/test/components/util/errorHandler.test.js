import React from 'react';
import { shallow } from 'enzyme';
import errorHandler from '../../../components/util/errorHandler';

describe('errorHandler', () => {
  let ErrorWrapper =
    errorHandler(() => <div />, 'Test error message', true, jest.fn());
  it('should return ErrorWrapper', () => {
    expect(ErrorWrapper).toBeDefined();
  });
  it('should match snapshot', () => {
    let wrapper = shallow(<ErrorWrapper />);
    expect(wrapper).toMatchSnapshot();
    wrapper = shallow(<ErrorWrapper error={new Error()} />);
    expect(wrapper).toMatchSnapshot();
    ErrorWrapper =
    errorHandler(() => <div />, 'Test error message');
    wrapper = shallow(<ErrorWrapper error={new Error()} />);
    expect(wrapper).toMatchSnapshot();
  });
});
