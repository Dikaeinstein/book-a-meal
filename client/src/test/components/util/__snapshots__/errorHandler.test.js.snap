// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorHandler should match snapshot 1`] = `<Component />`;

exports[`errorHandler should match snapshot 2`] = `
<div
  className="error-container text-center"
>
  <h4>
    Test error message
  </h4>
  <button
    className="btn btn-default"
    onClick={true}
    style={
      Object {
        "margin": "1rem auto 0 auto",
      }
    }
  >
    Retry
  </button>
</div>
`;

exports[`errorHandler should match snapshot 3`] = `
<div
  className="error-container text-center"
>
  <h4>
    Test error message
  </h4>
</div>
`;
